src/sql/src/normalize.rs

```diff
@@ -134,6 +134,17 @@ pub fn option_objects(options: &[SqlOption<Aug>]) -> BTreeMap<String, SqlOption<
         .collect()
 }
 
+pub fn with_options(options: &[WithOption]) -> BTreeMap<String, Value> {
```

Comment:
This is pretty opaque and oddly named--seems like you'd want to call it `with_option_value_values`, or something?

Also, why is this used? It skips over object names, but not totally clear to me to what end.

Reply:
This is intended to be the functional equivalent of `normalize::options` which works on `Vec<SqlOption<T>>` but I needed to transform a `Vec<WithOption>` instead and yields a map keyed on `WithOption.Key` with `WithOptionValue.Value` as the value.

Maybe this should be an `impl From<Vec<WithOption>> for BTreeMap` using some newtype wrapper on `Vec<WithOption>` to allow for the trait to be implemented?

Reply:
I think Sean is asking: why change the field type from `SqlOption<T>` to `WithOption`? The `CREATE SOURCE` path uses `SqlOption<T>`, so using `WithOption` here seems like asking for trouble.

Diff seems to go down cleanly.

```diff
diff --git a/src/sql-parser/src/ast/defs/ddl.rs b/src/sql-parser/src/ast/defs/ddl.rs
index 59a501777..3afd76099 100644
--- a/src/sql-parser/src/ast/defs/ddl.rs
+++ b/src/sql-parser/src/ast/defs/ddl.rs
@@ -458,14 +458,14 @@ impl_display!(DbzMode);
 
 #[derive(Debug, Clone, PartialEq, Eq, Hash, EnumKind)]
 #[enum_kind(ConnectorType)]
-pub enum CreateConnector {
+pub enum CreateConnector<T: AstInfo> {
     Kafka {
         broker: String,
-        with_options: Vec<WithOption>,
+        with_options: Vec<SqlOption<T>>,
     },
 }
 
-impl AstDisplay for CreateConnector {
+impl<T: AstInfo> AstDisplay for CreateConnector<T> {
     fn fmt<W: fmt::Write>(&self, f: &mut AstFormatter<W>) {
         match self {
             Self::Kafka {
@@ -481,7 +481,7 @@ impl AstDisplay for CreateConnector {
         }
     }
 }
-impl_display!(CreateConnector);
+impl_display_t!(CreateConnector);
 
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub enum KafkaConnector {
diff --git a/src/sql-parser/src/ast/defs/statement.rs b/src/sql-parser/src/ast/defs/statement.rs
index 14f112648..a75ccb40d 100644
--- a/src/sql-parser/src/ast/defs/statement.rs
+++ b/src/sql-parser/src/ast/defs/statement.rs
@@ -37,7 +37,7 @@ pub enum Statement<T: AstInfo> {
     Copy(CopyStatement<T>),
     Update(UpdateStatement<T>),
     Delete(DeleteStatement<T>),
-    CreateConnector(CreateConnectorStatement),
+    CreateConnector(CreateConnectorStatement<T>),
     CreateDatabase(CreateDatabaseStatement),
     CreateSchema(CreateSchemaStatement),
     CreateSource(CreateSourceStatement<T>),
@@ -383,13 +383,13 @@ impl_display!(CreateSchemaStatement);
 
 /// `CREATE CONNECTOR`
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
-pub struct CreateConnectorStatement {
+pub struct CreateConnectorStatement<T: AstInfo> {
     pub name: UnresolvedObjectName,
-    pub connector: CreateConnector,
+    pub connector: CreateConnector<T>,
     pub if_not_exists: bool,
 }
 
-impl AstDisplay for CreateConnectorStatement {
+impl<T: AstInfo> AstDisplay for CreateConnectorStatement<T> {
     fn fmt<W: fmt::Write>(&self, f: &mut AstFormatter<W>) {
         f.write_str("CREATE CONNECTOR ");
         if self.if_not_exists {
@@ -411,7 +411,7 @@ impl AstDisplay for CreateConnectorStatement {
         }
     }
 }
-impl_display!(CreateConnectorStatement);
+impl_display_t!(CreateConnectorStatement);
 
 /// `CREATE SOURCE`
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
diff --git a/src/sql-parser/src/parser.rs b/src/sql-parser/src/parser.rs
index 07e45d10d..e26ac47ea 100644
--- a/src/sql-parser/src/parser.rs
+++ b/src/sql-parser/src/parser.rs
@@ -1869,7 +1869,7 @@ impl<'a> Parser<'a> {
         let connector = if self.parse_keyword(KAFKA) {
             self.expect_keyword(BROKER)?;
             let broker = self.parse_literal_string()?;
-            let with_options = self.parse_opt_with_options()?;
+            let with_options = self.parse_opt_with_sql_options()?;
             CreateConnector::Kafka {
                 broker,
                 with_options,
diff --git a/src/sql/src/normalize.rs b/src/sql/src/normalize.rs
index d301e30cf..9b930eca3 100644
--- a/src/sql/src/normalize.rs
+++ b/src/sql/src/normalize.rs
@@ -28,7 +28,7 @@ use mz_sql_parser::ast::{
     CreateSinkStatement, CreateSourceStatement, CreateTableStatement, CreateTypeAs,
     CreateTypeStatement, CreateViewStatement, Function, FunctionArgs, Ident, IfExistsBehavior, Op,
     Query, SqlOption, Statement, TableFactor, TableFunction, UnresolvedObjectName,
-    UnresolvedSchemaName, Value, ViewDefinition, WithOption, WithOptionValue,
+    UnresolvedSchemaName, Value, ViewDefinition,
 };
 
 use crate::names::{
@@ -134,17 +134,6 @@ pub fn option_objects(options: &[SqlOption<Aug>]) -> BTreeMap<String, SqlOption<
         .collect()
 }
 
-pub fn with_options(options: &[WithOption]) -> BTreeMap<String, Value> {
-    options
-        .iter()
-        .filter_map(|o| match &o.value {
-            Some(WithOptionValue::Value(v)) => Some((o.key.to_string(), v.clone())),
-            Some(_) => None,
-            None => None,
-        })
-        .collect()
-}
-
 /// Unnormalizes an object name.
 ///
 /// This is the inverse of the [`unresolved_object_name`] function.
diff --git a/src/sql/src/plan/statement/ddl.rs b/src/sql/src/plan/statement/ddl.rs
index 34c55e46e..b3932964e 100644
--- a/src/sql/src/plan/statement/ddl.rs
+++ b/src/sql/src/plan/statement/ddl.rs
@@ -2657,14 +2657,14 @@ pub fn plan_create_secret(
 
 pub fn describe_create_connector(
     _: &StatementContext,
-    _: &CreateConnectorStatement,
+    _: &CreateConnectorStatement<Raw>,
 ) -> Result<StatementDesc, anyhow::Error> {
     Ok(StatementDesc::new(None))
 }
 
 pub fn plan_create_connector(
     scx: &StatementContext,
-    stmt: CreateConnectorStatement,
+    stmt: CreateConnectorStatement<Aug>,
 ) -> Result<Plan, anyhow::Error> {
     scx.require_experimental_mode("CREATE CONNECTOR")?;
 
@@ -2679,7 +2679,7 @@ pub fn plan_create_connector(
             broker,
             with_options,
         } => {
-            let mut with_options = normalize::with_options(&with_options);
+            let mut with_options = normalize::options(&with_options);
             ConnectorInner::Kafka {
                 broker: broker.parse()?,
                 config_options: kafka_util::extract_config(&mut with_options)?,
```

> Maybe this should be an impl From<Vec<WithOption>> for BTreeMap using some newtype wrapper on Vec<WithOption> to allow for the trait to be implemented?

Strikes me as moving in the wrong direction. Someone should figure out why we have both `WithOption` and `SqlOption<T>` when they both do basically the same thing. I'd love to delete one.

Reply:
In this case I couldn't find any options that were anything other than string literal values so `WithOption` seemed like the better fit as it avoided all of the type parameters and easily yielded correctly quoted and escaped values but since it appears only one other thing uses it I will switch back to `SqlOption<T>` since I do not recall what problem actually  made me dig into this and switch in the first place.

Reply:
Avoiding the complexity of type parameters doesn't seem like quite the right tradeoff here.

The `SqlOption`/`WithOption` bifurcation has been a thorn in our side, so I'll dedicate tomorrow to looking at it/resolving it, so I'd like to either move forward with Nikhil's suggestion for parameterizing this stuff, or wait until tomorrow/Monday to shake out what the solution would be.

Reply:
Yup, I originally went down this road because I was having some weird problem with the `SqlOption<T>` implementation I had and saw another statement with similar semantics using `WithOption` which did not require the specialization and it seemed to solve the problem at the time but using `SqlOption<T>` is working fine now so I have no idea what I was doing wrong.